Pre and post treatment of expression when cse=True in lambdify

mleila1312 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

### Fixes #26404
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
These additions to the lambdify file fixes the problems seen in the issue mentionned above.

### What is changed
When lambdify is called and the optionnal parameter cse==True, then we check if there are Derivatives in the arguments of the function. If so, they are replaced by variable names to prevent them from being ignored later in the cse process (see issue page [here](https://github.com/sympy/sympy/issues/26404) to see the problem). After this and the cse process is applied, we do a post process : we change back the Derivatives replaced at the beginning to their original values in the expression and in the cse replacements (ie the Derivatives).  

To do this, I have added two functions in lambdify.py : pre_treatment_cse(args, expr) (line 181 in lambdify.py) and post_treatment_cse(dictionnary, args, expr, cses)   (line 247 in lambdify.py) 

I have also added new tests in the test_lambdify.py file. (end of the file)

If you have any questions, would like more comments in the code.... , please let me know!


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * Fixed a bug with cse in lambdify when there are Derivatives in the arguments and the expression.

<!-- END RELEASE NOTES -->

Commit:
Fix cse treatment in lambdify with Derivatives

Before, when there were Derivatives in expr ans args given to lambdify
with cse enabled, there was an error because the cse treatment changed
the arguments of the Derivative object.

With this implementation, the expression is pre-treated by a function to
mask the instances of Derivative objects, then the cse process is
applied and finally we do a post-treatment to put back the Derivative
expressions in expr.

moorepants commented:
Hi, thanks for preparing this.

I have a couple of questions:

1. Is there a reason not to use the optimizations kwarg in cse for this (as I suggested in the issue)? The machinery is already there to to pre and post work on expression that are processed.

2. The pre and post functions seem elaborate, yet there is only one simple test. Do you need to test other possibilities given the complexity (I assume it is there there to handle other cases).

3. Also given the complexity mentioned in point 2., how much will this slow things down? You can use the `n_link_pendulum` model in the mechanics models module to generate some really long expressions with derivatives to test, if you need a test case.

mleila1312 commented:
@moorepants 
* concerning the optimization kwarg, I didn't use them here because I have some problems passing optionnal arguments to the function cse (you can see it in the last message on the issue, but basically, I noticed that I couldn't pass a flag that said cse is used in lambdify, it still retained the default value in the function, same for the optionnal argument list). So, by measure of security, I didn't want to implement it by passing the pre and post operations in the optionnal argument optimization. Although, with the way I implemented this, it would not be possible since a dictionnary containing the Derivative- newname associations is returned by the pre cse operation to be used in the post cse operation.
* concerning the test, I could add others. I can say that even if a Derivative (A) of a Derivative (B) and B are passed as arguments, it doesn't cause an issue, same if the Derivative is in a function but I could add them as a test also. (I will add it in my next commit, if you have other ideas for tests, don't hesitate)
* For the time complexity, if I test : 
```
import sympy as sm
import sympy.physics.mechanics.models as models
from sympy import (cos, sin, Matrix, symbols, zeros)                     
import sympy.physics.mechanics.tests as tests
from sympy.physics.mechanics import (dynamicsymbols)

def timing(func, a, b, c, d, e, f,g, h, i , j, k, l, m, n):
    import time
    s= time.time()
    print(func(a, b, c, d, e, f,g, h, i , j, k, l, m, n))
    print(time.time()- s)


l0, m0 = sm.symbols("l0 m0")
l1, m1 = sm.symbols("l1 m1")
m2 = sm.symbols("m2")
g = sm.symbols("g")
q0, q1, q2 = dynamicsymbols("q0 q1 q2")
u0, u1, u2 = dynamicsymbols("u0 u1 u2")
F, T1 = dynamicsymbols("F T1")
kane1 = models.n_link_pendulum_on_cart(2)
massmatrix1 = Matrix([[m0 + m1 + m2, -l0*m1*cos(q1) - l0*m2*cos(q1),
                           -l1*m2*cos(q2)],
                          [-l0*m1*cos(q1) - l0*m2*cos(q1), l0**2*m1 + l0**2*m2,
                           l0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2))],
                          [-l1*m2*cos(q2),
                           l0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2)),
                           l1**2*m2]])
timing(sm.lambdify((l0, m0 ,l1, m1, m2, g, q0, q1, q2 , u0, u1, u2, F, T1), massmatrix1 - \
            kane1.mass_matrix), 0, 0 ,1, 1, 2, 9, 0, 1, 2 , 0, 1, 2, 13, 1)
#and
timing(sm.lambdify((l0, m0 ,l1, m1, m2, g, q0, q1, q2 , u0, u1, u2, F, T1), massmatrix1 -\
            kane1.mass_matrix, cse=True), 0, 0 ,1, 1, 2, 9, 0, 1, 2 , 0, 1, 2, 13, 1)
```
For the not cse : 0.06650567054748535 s the first time, the second time -> 0.023397445678710938
For the one with cse= True : 0.06281447410583496 sthe first time, the second time -> 0.019282102584838867
So when there is no Derivative, it doesn't change anything

When there are Derivatives, I'm having trouble testing it, I tried : 
```
import sympy as sm
import sympy.physics.mechanics.models as models
from sympy import (cos, sin, Matrix, symbols, zeros)                     
import sympy.physics.mechanics.tests as tests
from sympy.physics.mechanics import (dynamicsymbols)
def timing(func, a, b, c, d, e, f,g, h, i , j, k, l, m, n):
    import time
    s= time.time()
    print(func(a, b, c, d, e, f,g, h, i , j, k, l, m, n))
    print(time.time()- s)

q0, q1, q2 = dynamicsymbols("q0 q1 q2")
u0, u1, u2 = dynamicsymbols("u0 u1 u2")
l0, m0 = sm.symbols("l0 m0")
l1, m1 = sm.Function('l1')(l0),sm.Function('m1')(m0)
m2 = l1.diff(l0)
Derivative(l1(l0), l0)
g = m2.diff(l0)
F, T1 = dynamicsymbols("F T1")
kane1 = models.n_link_pendulum_on_cart(2)
massmatrix1 = Matrix([[m0 + m1 + m2, -l0*m1*cos(q1) - l0*m2*cos(q1),
                           -l1*m2*cos(q2)],
                          [-l0*m1*cos(q1) - l0*m2*cos(q1), l0**2*m1 + l0**2*m2,
                           l0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2))],
                          [-l1*m2*cos(q2),
                           l0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2)),
                           l1**2*m2]])
timing(sm.lambdify((l0, m0 ,l1, m1, m2, g, q0, q1, q2 , u0, u1, u2, F, T1), massmatrix1 -\
                   kane1.mass_matrix), 0, 0 ,1, 1, 2, 9, 0, 1, 2 , 0, 1, 2, 13, 1)
```
I get the error :  
```
Traceback (most recent call last):
  File "<pyshell#139>", line 1, in <module>
    timing(sm.lambdify((l0, m0 ,l1, m1, m2, g, q0, q1, q2 , u0, u1, u2, F, T1), massmatrix1 - kane1.mass_matrix), 0, 0 ,1, 1, 2, 9, 0, 1, 2 , 0, 1, 2, 13, 1)
  File "<pyshell#127>", line 4, in timing
    print(func(a, b, c, d, e, f,g, h, i , j, k, l, m, n))
  File "<lambdifygenerated-1>", line 2, in _lambdifygenerated
    return ImmutableDenseMatrix([[_Dummy_56 + _Dummy_63 - m1 - m2, -_Dummy_56*l0*cos(_Dummy_61) - \
_Dummy_63*l0*cos(_Dummy_61) + l0*m1*cos(_Dummy_61) + l0*m2*cos(_Dummy_61), -\
_Dummy_56*_Dummy_64*cos(_Dummy_60) + l1*m2*cos(_Dummy_60)], [-_Dummy_56*l0*cos(_Dummy_61) - \
_Dummy_63*l0*cos(_Dummy_61) + l0*m1*cos(_Dummy_61) + l0*m2*cos(_Dummy_61), _Dummy_56*l0**2 + _Dummy_63*l0**2 -\
 l0**2*m1 - l0**2*m2, _Dummy_56*_Dummy_64*l0*(sin(_Dummy_60)*sin(_Dummy_61) + cos(_Dummy_60)*cos(_Dummy_61)) -\
 l0*l1*m2*(sin(_Dummy_60)*sin(_Dummy_61) + cos(_Dummy_60)*cos(_Dummy_61))], [-\
_Dummy_56*_Dummy_64*cos(_Dummy_60) + l1*m2*cos(_Dummy_60), _Dummy_56*_Dummy_64*l0*\
(sin(_Dummy_60)*sin(_Dummy_61) + cos(_Dummy_60)*cos(_Dummy_61)) - l0*l1*m2*(sin(_Dummy_60)*sin(_Dummy_61) +\
 cos(_Dummy_60)*cos(_Dummy_61)), _Dummy_64**2*_Dummy_56 - l1**2*m2]])
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python312\Lib\site-packages\mpmath\matrices\matrices.py", line 302, in __init__
    self[i, j] = a
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python312\Lib\site-packages\mpmath\matrices\matrices.py", line 560, in __setitem__
    value = self.ctx.convert(value)
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python312\Lib\site-packages\mpmath\ctx_mp_python.py", line 671, in convert
    return ctx._convert_fallback(x, strings)
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python312\Lib\site-packages\mpmath\ctx_mp.py", line 634, in _convert_fallback
    raise TypeError("cannot create mpf from " + repr(x))
TypeError: cannot create mpf from -m1 - m2 + 3
```

It seems that the derivatives aren't dummyfied in the matrix, this problem isn't a result of my changes I think since my changes are only applied when cse is True. Maybe I'm not using this right, otherwise, this could be another issue.

smichr commented:
You might consider using `_masked` from `solveset.py`.

mleila1312 commented:
>You might consider using `_masked` from solveset.py.

I looked into it, but I'm not sure if it would be compatible for dummyfying different derivatives, from what I've seen, it can only replace atoms like x, y, ... And a derivative has atoms as arguments, so I don't think it's an atom, but please tell me if I'm wrong, it would be more practical  

I've detected another problem, cses can also replace combinations of the replaced Derivatives in the pre-treatment, I've solved this issue, I'm in the process of building a test with some matrix, I will push the updates soon.

mleila1312 commented:
About the time of execution, in this new case(with more arguments ans more instances of Derivatives to replace  in the expression) :  
```
l0, m0 = sm.symbols("l0 m0")
l1, m1 = sm.symbols("l1 m1")
m2 = sm.symbols("m2")
g = symbols("g")
q0, q1, q2 = sm.Function("q0")(l0),sm.Function("q1")(l1),sm.Function("q2")(m0)
u1, u2 =q1.diff(l1), q2.diff(m0)
F, T1 = dynamicsymbols("F T1")
massmatrix1 = Matrix([[m0 + m1 + m2, -l0*m1*cos(q1) - l0*m2*cos(q1),
                           -l1*m2*cos(q2)],
                          [-l0*m1*cos(q1) - l0*m2*cos(q1), l0**2*m1 + l0**2*m2,
                           l0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2))],
                          [-l1*m2*cos(q2),
                           l0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2)),
                           l1**2*m2]])

forcing1 = Matrix([[-l0*m1*u1**2*sin(q1) - l0*m2*u1**2*sin(q1) -
                        l1*m2*u2**2*sin(q2) + F,
                       g*l0*m1*sin(q1) + g*l0*m2*sin(q1) -
                        l0*l1*m2*(sin(q1)*cos(q2) - sin(q2)*cos(q1))*u2**2,
                       g*l1*m2*sin(q2) - l0*l1*m2*(-sin(q1)*cos(q2) +
                                                    sin(q2)*cos(q1))*u1**2],
                   [-l0*m1*u1**2*sin(q1) - l0*m2*u1**2*sin(q1) -
                        l1*m2*u2**2*sin(q2) + F,
                       g*l0*m1*sin(q1) + g*l0*m2*sin(q1) -
                        l0*l1*m2*(sin(q1)*cos(q2) - sin(q2)*cos(q1))*u2**2,
                       g*l1*m2*sin(q2) - l0*l1*m2*(-sin(q1)*cos(q2) +
                                                    sin(q2)*cos(q1))*u1**2],
                   [-l0*m1*u1**2*sin(q1) - l0*m2*u1**2*sin(q1) -
                        l1*m2*u2**2*sin(q2) + F,
                       g*l0*m1*sin(q1) + g*l0*m2*sin(q1) -
                        l0*l1*m2*(sin(q1)*cos(q2) - sin(q2)*cos(q1))*u2**2,
                       g*l1*m2*sin(q2) - l0*l1*m2*(-sin(q1)*cos(q2) +
                                                    sin(q2)*cos(q1))*u1**2]])
# time of this =t1
(lambdify((l0, m0 ,l1, m1, m2, g, q0, q1, q2 ,  u1, u2, F, T1), massmatrix1 -\
           forcing1)( 0, 0 ,1, 1, 2, 9, 0, 1, 2 , 1, 2, 13, 1))
#time of this = t2
(lambdify((l0, m0 ,l1, m1, m2, g, q0, q1, q2 ,  u1, u2, F, T1), massmatrix1 -forcing1, \
                  cse=True)( 0, 0 ,1, 1, 2, 9, 0, 1, 2 , 1, 2, 13, 1))
```
t1 = 0.04956769943237305s on first try, 0.052608489990234375s on second try  
t2 = 0.03235220909118652s on first try, 0.04843401908874512s on second try  

If you want me to do more complicated tests, please let me know

abhiphile commented on .mailmap:
```diff
@@ -1546,6 +1546,7 @@ anutosh491 <andersonbhat491@gmail.com> Anutosh Bhat <87052487+anutosh491@users.n
 anutosh491 <andersonbhat491@gmail.com> anutosh491 <87052487+anutosh491@users.noreply.github.com>
 arooshiverma <av22@iitbbs.ac.in>
 atharvParlikar <atharvparlikar@gmail.com>
+bin/mailmap_check.py
```
You don't need to include this in `.mailmap`.

mleila1312 replied:
yes, you're right, sorry I must have put it there by mistake (I'll correct it), thank you!

abhiphile commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+    # creation of the symbols that can't be used to replace the Derivatives in the expression
+    excluded_symbols = set()
+    symbols = numbered_symbols(cls=Symbol)
+    def _eliminates_symbols(expr) :
```
```suggestion
    def _eliminates_symbols(expr):
```

mleila1312 replied:
yes, you're right, sorry I must have put it there by mistake, thank you!

abhiphile commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+        list(map(_eliminates_symbols, args))
+        return
+    # gets all the symbols that can't be used
+    if iterable(expr) :
```
```suggestion
    if iterable(expr):
```

abhiphile commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+    from sympy.core.function import Derivative
+    post_expr=expr
+    post_cses=cses
+    for arg in args :
```
Same here 
```suggestion
    for arg in args:
```

abhiphile commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -1891,3 +1891,64 @@ def test_assoc_legendre_numerical_evaluation():
+    x = Function("x")(t)
+    xd = x.diff(t)
+    xdd= xd.diff(t)
+    assert lambdify((xd, x), xd, cse=True)(1, 1) ==1
```
```suggestion
    assert lambdify((xd, x), xd, cse=True)(1, 1) == 1
```

abhiphile commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -1891,3 +1891,64 @@ def test_assoc_legendre_numerical_evaluation():
+    xd = x.diff(t)
+    xdd= xd.diff(t)
+    assert lambdify((xd, x), xd, cse=True)(1, 1) ==1
+    assert lambdify((xd, x), xd + x, cse=True)(1, 1) ==2
```
Same for all.
```suggestion
    assert lambdify((xd, x), xd + x, cse=True)(1, 1) == 2
```

mleila1312 replied:
yes you're right, I'll put it in the next commit too!

abhiphile commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -1891,3 +1891,64 @@ def test_assoc_legendre_numerical_evaluation():
+                        g*l1*m2*sin(q2) - x0*l1*m2*(-sin(q1)*cos(q2) +
+                                                        sin(q2)*cos(q1))*u1**2]])
+    res_expected=Matrix([[ 1., 0, 0],[-1.,  0, 0],[-1.,  0,  0]])
+    res_lamdbify=Matrix((lambdify((x0, m0 ,l1, m1, m2, g, q0, q1, q2 ,  u1, u2, F, T1), massmatrix1 -forcing1, \
```
```suggestion
    res_lamdbify=Matrix((lambdify((x0, m0 ,l1, m1, m2, g, q0, q1, q2, u1, u2, F, T1), massmatrix1 -forcing1, \
```

abhiphile commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+            for i in range(len(cses)) :
+                new_a,a = cses[i]
+                if a.has(dictionnary[arg]):
+                    if a==dictionnary[arg]:
```
Same for all
```suggestion
                    if a == dictionnary[arg]:
```

abhiphile commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+                _eliminates_symbols(expr)
+     #gets the possible symbols to replace Derivatives with
+    symbols = (_ for _ in symbols if _.name not in excluded_symbols)
+    new_expr=expr
```
```suggestion
    new_expr = expr
```

abhiphile commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
 _lambdify_generated_counter = 1
 
+def pre_treatment_cse(args_f, expr):
+    r'''
```
```suggestion
    r"""
```

mleila1312 replied:
corrected this also in pre treatment and post treatment

moorepants commented:
To test the effects of slowdown you need something like `kane1 = models.n_link_pendulum_on_cart(20)` at least.

mleila1312 commented:
> To test the effects of slowdown you need something like `kane1 = models.n_link_pendulum_on_cart(20)` at least.

I've done several tests to see the duration of execution.
For `kane = models.n_link_pendulum_on_cart(20)` , average for a100 executions : 
* without cse : 1.0021281242370604 ms
* with cse : 0.5107760429382324 ms

Here, activating cse reduces the time of execution  by half.

For `kane = models.n_link_pendulum_on_cart(40)` , average for a 100 executions : 
* without cse : 5.916361808776856 ms
* with cse : 2.1598124504089355 ms

Here, activating cse reduces the time of execution  by 63%.

For `kane = models.n_link_pendulum_on_cart(100)` , average for a 100 executions : 
* without cse : 81.51222705841064 ms
* with cse : 22.987971305847166 ms

Here, activating cse reduces the time of execution  by 72%.

moorepants commented:
Sorry if I misstated. The slowdown I'm interested in is the timing of lambdify(). If we do pre and post substitutions, there is a slowdown associated with executing the lambdify function. It will only be seen on very large expressions that are full of `Derivative()`  terms.

moorepants commented:
Are you saying that your modification makes lambdify faster?

moorepants commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+    from sympy.matrices.expressions import MatrixSymbol
+    from sympy.matrices.expressions.matexpr import MatrixElement
+    from sympy.polys.rootoftools import RootOf
+    from sympy.utilities.iterables import numbered_symbols, iterable
```
These should all go to the top of the file unless there is a circular import dependency issue.

mleila1312 replied:
Only from `sympy.utilities.iterables import numbered_symbols, iterable` didn't cause a circular import dependency issue, I put it at the top of the file(I'll commit in one commit)

moorepants commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -1891,3 +1891,64 @@ def test_assoc_legendre_numerical_evaluation():
+                            x0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2))],
+                            [-l1*m2*cos(q2),
+                            x0*l1*m2*(sin(q1)*sin(q2) + cos(q1)*cos(q2)),
+                            l1**2*m2]])
```
These expressions would be more appropriate if the `u` values were all swapped with derivatives: `u1 = Derivative(q1(t) t)`, for example.

mleila1312 replied:
u values are : u1, u2 =`q1.diff(l1), q2.diff(m0)`, so already derivatives, did you mean putting more of them in the expression of the matrix? Or do you mean putting Derivatives directly in the matrix? If so, I haven't handled the case were the expression of a Derivative is directly in the expression and not also as an argument, I will add it

Mohamad1999hadi replied:
@moorepants @mleila1312 @sympy-bot @moorepants @tjl @scolobb @abhiphile @xrmx you are the best and you deserve the world to be better for your kids because I am proud to say I have been a teacher and teacher and a great friend and I will continue working with my kids to help you through the process and make [_****_]()

moorepants commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
 # linecache.
 _lambdify_generated_counter = 1
 
+def pre_treatment_cse(args_f, expr):
```
These functions should probably be private: append an underscore to the name: `_pre_treatment_cse()`.

moorepants commented on sympy/utilities/lambdify.py:
```diff
@@ -178,6 +178,102 @@ def _import(module, reload=False):
+    This function returns the new expressions with the instances of Derivatives replaced
+    and a dictionnary that contains the associations of changes
+    (ex : Derivative(x, u) : "x0")
+    """
```
Please follow the docstring style guide: https://docs.sympy.org/latest/contributing/docstring.html

moorepants commented on sympy/utilities/lambdify.py:
```diff
@@ -866,8 +961,13 @@ def _lambdifygenerated(x):
-        cses, _expr = _cse(expr, list=False)
+        #get the dictionnary containing the Derivative in the
+        #arguments and their new name
+        dictionnary, new_expr= pre_treatment_cse(args, expr)
```
```suggestion
        dictionary, new_expr= pre_treatment_cse(args, expr)
```

mleila1312 commented:
>Sorry if I misstated. The slowdown I'm interested in is the timing of lambdify(). If we do pre and post substitutions, there is a slowdown associated with executing the lambdify function. It will only be seen on very large expressions that are full of Derivative() terms.

No problem, I wasn't sure sorry! I'll do that today(so max in 8 hours) and post the new tests results, but I won't be able to compare to before the implementation(since it didn't work, although I could compare to the execution time of expressions without Derivative terms)

> Are you saying that your modification makes lambdify faster?

It seems counterintuitive, I still wanted to add the results  since the goal was to compare, but the fact that the execution is faster with the implementation might be due to other processes that were running on my computer (I didn't execute anything else, but I don't see any other explanation since the kane matrix didn't have any Derivative so didn't need modifications by the pre treatment)

moorepants commented:
Ideally a fix would work with any types of SymPy objects where this may occur. `Derivative()` is one, but this may also occur with things like `Integral(f(t), t)` or even things like `x(t) + sin(t)`. Is your solution only for Derivative()?

moorepants commented:
> No problem, I wasn't sure sorry! I'll do that today(so max in 8 hours) 

No apologies are needed because you have not done anything wrong and there are no time limits so please take whatever time you need/desire.

mleila1312 commented:
> Derivative() is one, but this may also occur with things like Integral(f(t), t) or even things like x(t) + sin(t). Is your solution only for Derivative()?

```
l0=symbols('l0')
x = Function('x')(sin(l0)+l0)
y=Function('y')(x.diff(l0)*x.diff(l0))
sm.lambdify((y, x, l0), y+x*y*l0 +x*x +l0*x*x, cse=True)(1,1,1)
```
And
```
In[1] : print(y+x*y*lo +x*x +lo*x*x)
                                         
Out[1] : lo*x(lo + sin(lo))**2 + 
             lo*x(lo + sin(lo))*y((cos(lo) + 1)**2*Subs(Derivative(x(_xi_1), _xi_1), _xi_1, lo + sin(lo))**2) +
             x(lo + sin(lo))**2 + y((cos(lo) + 1)**2*Subs(Derivative(x(_xi_1), _xi_1), _xi_1, lo + sin(lo))**2)
```

With this, I get the original error, so ideally, a solution for this case should be implemented as well. I checked and the pre and post operations did not change the expression.  And 
```
sm.lambdify((x, lo), x*lo +x*x +lo*x*x, cse=True)(1,1)
```
Worked correctly

For other classes of objects, I think my solution would work, we would just need to add to the isinstance line 238 in lambdify.py all the other classes we want to mask, but it should be tested. 

Although, concerning the integral, I tried to pass it as an argument (without cse activated) and I got an error, but maybe I didn't write the command lines correctly.

> No apologies are needed because you have not done anything wrong and there are no time limits so please take whatever time you need/desire.

Thank you

mleila1312 commented:
I fixed an new issue(list of expressions given, not just one)

I did the new tests, here are the results on average for a 100 executions : 

* for kane 20 : 
      + with cse enabled : 10.001659393310548 μs
      + without cse : 9.942054748535157 μs

Execution with cse is 0.006% shorter.

* for kane 40 : 
      + with cse enabled : 49.996376037597654 μs
      + without cse : 49.99399185180664 μs

Execution with cse is 0.00004% shorter.

* for kane 100 :
      + with cse enabled : 0.8887672424316407 ms
      + without cse : 0.892937183380127 ms

Execution with cse is 0.004% shorter.

The percentage doesn't mean much since it's really small but if you want to execute yourselves the tests I've done and measure the average time spent per execution, they are [here
](https://github.com/mleila1312/test_files/blob/main/test_derivative_issue26404.py)  

I changed and measured the time of execution with different Matrices(or list of lists here) similar to the kane built-in(the built-in function didn't take Derivatives as arguments so I made one similar in dimensions)

mleila1312 commented:
Rebased with more explicit commit message
